ref(consumers): Allow injecting StreamProcessor kwargs using options

untitaker opened the pull request:
We have had situations where applying processor arguments using CLI
arguments was error prone and took a long time to revert if something is
wrong. Let's use dynamic cli flags instead.

currently you have to:
* add a feature to arroyo
* add a cli flag to sentry
* add cli in k8s

then the flag cleanup is all of that in reverse

with this pr you at least only have to bump arroyo in sentry, and there's nothing to be done in sentry on cleanup

untitaker committed:
ref(consumers): Allow injecting StreamProcessor kwargs using options

We have had situations where applying processor arguments using CLI
arguments was error prone and took a long time to revert if something is
wrong. Let's use options instead. We can now set `join_timeout`,
`stuck_detector_timeout` etc using options, roll those settings out
gradually and eventually set the defaults in code.

untitaker commented on src/sentry/consumers/__init__.py:
```diff
@@ -33,6 +33,39 @@
+    consumer_specific_overrides = overrides.get(consumer_name, {})
+
+    # Apply overrides
+    for key, value in consumer_specific_overrides.items():
```
sure, if we set a bad option value it'll fall apart. that's a big risk but i don't know how to guard against that. if we had a dynamic settings map specifically supported in arroyo with proper validation, it would be safer but also more complex in its own ways.

untitaker replied:
@evanh if we did that validation in sentry then we would have to once again make a PR against sentry everytime we add a new option.

victoria-yining-huang commented:
i think this in general is a good idea, I just have one concern and would like to know if it's considered already:
- how to make sure eventually the setting is set in sentry repo and not just forgotten in the options repo? 

if we are only counting on people to stick to "best practices" this will eventually lead to things living in two places, which I'm not a fan of.

bmckerry reviewed (commented):
> and eventually set the defaults in code

My concern is that this won't happen. We've seen this before with Sentry regionsilo config options in the [ops repo override](https://github.com/getsentry/ops/blob/master/k8s/services/getsentry/templates/k8s_regionsilo_ops.py.j2#L10-L14) being forgotten about and never making it into our actual regionsilo config file.
If we want to avoid accidentally killing a consumer by deploying a flag in ops that doesn't exist yet, instead of setting CLI flags thru options, could we have both sentry's `get_stream_processor()` and the arroyo `StreamProcessor` take an arbitrary `**kwargs` that it does nothing with?

Correct me if I'm wrong, but if this works how I think that it does then if we pass an arg in the ops repo that Sentry consumer code isn't able to handle yet, it's a no-op. If we add that specific arg to `get_stream_processor()` above the `**kwargs` then we can handle that arg in Sentry and pass it on to StreamProcessor.

untitaker commented on src/sentry/consumers/__init__.py:
```diff
@@ -623,15 +656,26 @@ def build_consumer_config(group_id: str):
+        consumer_name, processor_args, processor_args_overrides
     )
 
+    return StreamProcessor(**processor_args)
```
yeah. we can try to handle that case

untitaker committed:
skip invalid parameters in consumers

untitaker committed:
Merge branch 'master' into ref/processor-args

untitaker commented on src/sentry/consumers/__init__.py:
```diff
@@ -623,15 +676,26 @@ def build_consumer_config(group_id: str):
+    processor_args_overrides = options.get("consumer.arroyo.processor_args")
+    processor_args = apply_processor_args_overrides(
+        consumer_name, processor_args, processor_args_overrides
     )
```
options are just way faster to apply than any of those other solutions. faster to apply, faster to revert.

untitaker committed:
move to cli

untitaker committed:
Merge remote-tracking branch 'origin/master' into ref/processor-args

untitaker committed:
wip

untitaker committed:
wip

untitaker committed:
fix(consumers): Fix type extraction for Union types in CLI args

The previous logic checked `get_origin(param_type) is not None` which
incorrectly matched ANY parameterized generic type (like List[int],
Dict[str, Any]), not just Optional/Union types. This caused the code
to extract the wrong type for non-Union generics.

Now specifically checking `get_origin(param_type) is Union` to only
extract types from Optional[T] (which is Union[T, None]).

Addresses Cursor review feedback.
